src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.cs

```diff
@@ -1785,6 +1825,47 @@ public DagState(ImmutableArray<StateForCase> cases, ImmutableDictionary<BoundDag
             // After the entire graph of DagState objects is complete, we translate each into its Dag node.
             public BoundDecisionDagNode? Dag;
 
+            private DagState()
+            {
+            }
+
+            public void ClearAndFree()
+            {
+                Cases.Free();
+                Cases = default;
+                RemainingValues = null!;
+                SelectedTest = null;
+                TrueBranch = null;
+                FalseBranch = null;
+                Dag = null;
+
+                s_dagStatePool.Free(this);
+            }
+
+            /// <summary>
+            /// Created an instance of <see cref="DagState"/>.  Will take ownership of <paramref name="cases"/>.  That
+            /// <see cref="ArrayBuilder{StateForCase}"/> will be returned to its pool when <see cref="ClearAndFree"/> is
+            /// called on this.
+            /// </summary>
+            public static DagState Create(ArrayBuilder<StateForCase> cases, ImmutableDictionary<BoundDagTemp, IValueSet> remainingValues)
```

Comment:
>ArrayBuilder<StateForCase> cases [](http://example.com/codeflow?start=42&length=32)

nit: the value of freezing is that the caller has confidence that what is passed in will be hard to mutate accidentally. Would it be better then to make the caller responsible for freezing? (ie. change the type of `cases` parameter to `FrozenArrayBuilder<StateForCase>`)

Reply:
wfm.

Reply:
Done.  That's a lot nicer.